### PR TITLE
Check namespaces when resolving associated items in typeck

### LINKED ISSUE #35600: Associated types with the same name as a function make an unimplementable trait
``` rust
trait Foo {
    type bar;
    fn bar();
}
```

is accepted (though generally considered bad style), but is unimplementable:

``` rust
impl Foo for () {
    type bar = ();
    fn bar() {
    }
}
```

gives the error "item `bar` is an associated method, which doesn't match its trait `<() as Foo>`"

This seems kinda silly. Allowing this non-conflicting name reuse might be useful for macros as long as procedural macros don't yet exist and concat_idents is useless. Forbidding it outright seems fine, but should be done at the trait definition.

### LINKED ISSUE #44247: Name resolution does not work if associated type and constant use the same name
Consider this program:

```Rust
trait T {
    type X;
    const X: Self::X;
}
fn foo<X: T>() {
    let _: X::X = X::X;
}
```

I believe it should compile OK, but currently (Rust 1.20) it fails to compile with a wrong error message:

<details>

<summary>Error</summary>

```
   Compiling example_project v0.1.0 (file:///home/matklad/projects/intellij-rust/exampleProject)
error[E0599]: no associated item named `X` found for type `X` in the current scope
 --> src/main.rs:6:19
  |
6 |     let _: X::X = X::X;
  |                   ^^^^
  |
  = note: found the following associated functions; to be used as methods, functions must have a `self` parameter
note: candidate #1 is defined in the trait `T`
 --> src/main.rs:2:5
  |
2 |     type X;
  |     ^^^^^^^
  = help: to disambiguate the method call, write `T::X(...)` instead
  = help: items from traits can only be used if the trait is implemented and in scope
  = note: the following trait defines an item `X`, perhaps you need to implement it:
          candidate #1: `T`

error: aborting due to previous error

error: Could not compile `example_project`.

To learn more, run the command again with --verbose.
```

</details>

For comparison, the following program builds just fine:

```Rust
mod X {
    pub type X = i32;
    pub const X: self::X = 0;
}
fn foo() {
    let _: X::X = X::X;
}
```

### PR DESCRIPTION
**Pull request by matthewjasper**
Closes #35600
Closes #44247 
Fixes a "cannot move a value of type..." error in the same case as #44247 but with the associated items swapped.

### CODE REVIEW

File: src/librustc_typeck/check/method/probe.rs [hunk 1]
```diff
@@ -1317,11 +1318,12 @@ impl<'a, 'gcx, 'tcx> ProbeContext<'a, 'gcx, 'tcx> {
...
                     .filter(|x| {
                         let dist = lev_distance(&*name.as_str(), &x.name.as_str());
-                        dist > 0 && dist <= max_dist
+                        Namespace::from(x.kind) == Namespace::Value && dist > 0
+                        && dist <= max_dist
                     })
                     .collect()
             } else {
-                self.fcx.associated_item(def_id, name).map_or(Vec::new(), |x| vec![x])
+                self.fcx.associated_item(def_id, name, Namespace::Value).map_or(Vec::new(), |x| vec![x])
```
Comment (kennytm) [inline src/librustc_typeck/check/method/probe.rs]: This line is too long.

```
[00:03:58] tidy error: /checkout/src/librustc_typeck/check/method/probe.rs:1326: line longer than 100 chars
```

File: src/librustc_typeck/namespace.rs [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+// except according to those terms.
+
+// Whether an item exists in the type or value namespace.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+pub enum Namespace {
+    Type,
+    Value,
+}
+
+impl From<ty::AssociatedKind> for Namespace {
```
Comment (kennytm) [inline src/librustc_typeck/namespace.rs]: You need to import the module `ty`…

```rust
use rustc::ty;
```

File: src/librustc_typeck/coherence/inherent_impls_overlap.rs [hunk 3]
```diff
@@ -8,6 +8,7 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
+use namespace::Namespace;
```
Comment (kennytm) [inline src/librustc_typeck/coherence/inherent_impls_overlap.rs]:
```
[00:12:43] error: unused import: `self`
[00:12:43]   --> /checkout/src/librustc_typeck/coherence/inherent_impls_overlap.rs:16:17
[00:12:43]    |
[00:12:43] 16 | use rustc::ty::{self, TyCtxt};
[00:12:43]    |                 ^^^^
[00:12:43]    |
```

**Review by kennytm — APPROVED** [review]

**Comment by petrochenkov** [comment]
This is great, two less issues on my backlog! :)
@bors r+

**Pull request merged**